### PR TITLE
Fix NumPy broadcasting in lstsq_backward

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91460
* #91456

Fixes https://github.com/pytorch/pytorch/issues/77225

cc @albanD

### CODE REVIEW

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -3661,37 +3661,38 @@ Tensor linalg_lstsq_jvp(
...
-    // B_grad = std::get<0>(at::linalg_lstsq(A.transpose(-1, -2).conj(), grad,
-    // rcond, driver)); but we avoid this approach as `gelsy` is
-    // non-deterministic
-    B_grad = pinvA.transpose(-1, -2).conj().matmul(grad);
+    // B_grad = std::get<0>(at::linalg_lstsq(A.mH(), gX, rcond, driver));
+    // but we avoid this approach as `gelsy` is non-deterministic
+    B_grad = matrix_to_vector(pinvA.mH().matmul(gX));
   }
 
   return std::make_tuple(A_grad, B_grad);
```
Comment (Skylion007) [inline torch/csrc/autograd/FunctionsManual.cpp:3698]:
```suggestion
  return std::make_tuple(std::move(A_grad), std::move(B_grad));
```

Comment (lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:3698]: I mean, I don't think it's too important to do those kind of optimisations in an operation thats computing an SVD and 10+ matmuls really. I think that readability is more important here.

**Comment by lezcano** [comment]
The errors in both PRs are unrelated.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -3661,37 +3661,38 @@ Tensor linalg_lstsq_jvp(
...
-    const Tensor& B,
-    const c10::optional<double> rcond,
-    const c10::optional<c10::string_view> driver,
-    const std::array<bool, 2>& grad_input_mask) {
+    const Tensor& B_) {
   at::NoTF32Guard disable_tf32;
-  Tensor A_grad, B_grad;
-  if (!grad.defined()) {
-    return std::make_tuple(A_grad, B_grad);
+  auto A_requires_grad = A.requires_grad();
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:3668]: You should use `grad_input_mask` that is provided to the formula to decide what to do. Not the requires_grad field.

**Review by albanD — COMMENTED** [review]
That sounds ok to me

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 2 above)
Comment (albanD) [reply torch/csrc/autograd/FunctionsManual.cpp:3668]: This is important for cases where user uses `autograd.grad()` or `backward(inputs=)`: some Tensors might require gradients but their gradient is not needed for this particular call to backward.

Comment (lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:3668]: ahhh, fair. We should fix that in many other linalg formulas then!

**Review by albanD — APPROVED** [review]
SGTM
